subsys/pcd/src/pcd.c

```diff
@@ -0,0 +1,156 @@
+#include <pm_config.h>
+#include <hal/nrf_reset.h>
+#include <hal/nrf_spu.h>
+/** Offset which the application should be copied into */
+#define NET_CORE_APP_OFFSET PM_CPUNET_B0N_SIZE
+#endif
+
+struct pcd_cmd {
+	uint32_t magic; /* Magic value to identify this structure in memory */
+	const void *data;     /* Data to copy*/
+	size_t len;           /* Number of bytes to copy */
+	off_t offset;         /* Offset to store the flash image in */
+} __aligned(4);
+
+static struct pcd_cmd *cmd = (struct pcd_cmd *)PCD_CMD_ADDRESS;
+
+void pcd_fw_copy_invalidate(void)
+{
+	cmd->magic = PCD_CMD_MAGIC_FAIL;
+}
+
+enum pcd_status pcd_fw_copy_status_get(void)
+{
+	if (cmd->magic == PCD_CMD_MAGIC_COPY) {
+		return PCD_STATUS_COPY;
+	} else if (cmd->magic == PCD_CMD_MAGIC_DONE) {
+		return PCD_STATUS_COPY_DONE;
+	}
+
+	return PCD_STATUS_COPY_FAILED;
+}
+
+int pcd_fw_copy(struct device *fdev)
+{
+	struct stream_flash_ctx stream;
+	uint8_t buf[CONFIG_PCD_BUF_SIZE];
+	int rc;
+
+	rc = stream_flash_init(&stream, fdev, buf, sizeof(buf),
+			       cmd->offset, 0, NULL);
+	if (rc != 0) {
+		LOG_ERR("stream_flash_init failed: %d", rc);
+		return rc;
+	}
+
+	rc = stream_flash_buffered_write(&stream, (uint8_t *)cmd->data,
+					 cmd->len, true);
+	if (rc != 0) {
+		LOG_ERR("stream_flash_buffered_write fail: %d", rc);
+		return rc;
+	}
+
+	LOG_INF("Transfer done");
+
+	return 0;
+}
+
+void pcd_fw_copy_done(void)
+{
+	/* Signal complete by setting magic to DONE */
+	cmd->magic = PCD_CMD_MAGIC_DONE;
+}
+
+#if defined(CONFIG_SOC_NRF5340_CPUAPP) && defined(CONFIG_MCUBOOT)
+
+/** @brief Construct a PCD CMD for copying data/firmware.
+ *
+ * @param data   The data to copy.
+ * @param len    The number of bytes that should be copied.
+ * @param offset The offset within the flash device to write the data to.
+ *               For internal flash, the offset is the same as the address.
+ *
+ * @retval non-negative integer on success, negative errno code on failure.
+ */
+static int pcd_cmd_write(const void *data, size_t len, off_t offset)
+{
+	if (data == NULL || len == 0) {
+		return -EINVAL;
+	}
+
+	cmd->magic = PCD_CMD_MAGIC_COPY;
+	cmd->data = data;
+	cmd->len = len;
+	cmd->offset = offset;
+
+	return 0;
+}
+
+int pcd_network_core_update(const void *src_addr, size_t len)
+{
+	int err;
+
+	/* Retain nRF5340 Network MCU in Secure domain (bus
+	 * accesses by Network MCU will have Secure attribute set).
+	 * This is needed for the network core to be able to read the
+	 * shared RAM area used for IPC.
+	 */
+	nrf_spu_extdomain_set(NRF_SPU, 0, true, false);
+
+	/* Ensure that the network core is turned off */
+	nrf_reset_network_force_off(NRF_RESET, true);
+
+	err = pcd_cmd_write(src_addr, len, NET_CORE_APP_OFFSET);
+	if (err != 0) {
+		LOG_INF("Error while writing PCD cmd: %d", err);
+		return err;
+	}
+
+	nrf_reset_network_force_off(NRF_RESET, false);
+	LOG_INF("Turned on network core");
+
+	do {
+		err = pcd_fw_copy_status_get();
```

Comment:
We could save some power here by only checking every millisecond or so. I don't know how important it is though.

Reply:
IMO the added failure option is not worth the effort, I want to keep this code dead simple, but I can add a task to see if there is a simple solution to it.

Reply:
Task is fine. It might be as simple as `k_msleep(1)`